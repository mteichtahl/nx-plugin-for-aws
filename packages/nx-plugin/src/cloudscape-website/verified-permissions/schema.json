{
  "$schema": "https://json-schema.org/schema",
  "$id": "Verified permissions",
  "cli": "nx",
  "title": "AWS Verified Permissions for Cloudscape website.",
  "description": "Add AWS Verified Permissions to an existing ts#cloudscape-website#cognitoAuth",
  "type": "object",
  "properties": {
    "project": {
      "description": "The root directory of the Cloudscape application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "x-prompt": "The root directory for your Cloudscape application",
      "x-dropdown": "projects"
    },
    "namespace": {
      "description": "Namespace of the verified permissions schema. Can not be aws, amazon, or cedar.",
      "type": "string",
      "x-prompt": "What is the namespace for your verified permissions?",
      "x-priority": "important",
      "default": "Default"
    },
    "principalEntity": {
      "description": "The principal entity type that you want to associate with your identity source.",
      "type": "string",
      "x-prompt": "What is your principal entity?",
      "x-priority": "important",
      "default": "User"
    },
    "groupEntity": {
      "description": "The group entity type that you want to associate with your groups.",
      "type": "string",
      "x-prompt": "What is your group entity?",
      "x-priority": "important",
      "default": "Group"
    }
  },
  "required": ["project"]
}
