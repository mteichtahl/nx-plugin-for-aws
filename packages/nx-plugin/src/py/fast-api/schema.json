{
  "$schema": "https://json-schema.org/schema",
  "$id": "PyFastApi",
  "title": "Create a Python FastAPI project",
  "description": "Create a Python FastAPI project",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "Name of the API project to generate",
      "x-priority": "important",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like your API to have? i.e: MyApi"
    },
    "computeType": {
      "type": "string",
      "description": "The type of compute to use to deploy this API. Choose between ServerlessApiGatewayRestApi (default) or ServerlessApiGatewayHttpApi.",
      "default": "ServerlessApiGatewayRestApi",
      "enum": ["ServerlessApiGatewayRestApi", "ServerlessApiGatewayHttpApi"],
      "x-prompt": "What compute type would you like to deploy your API with?",
      "x-priority": "important"
    },
    "auth": {
      "type": "string",
      "description": "The method used to authenticate with your API. Choose between IAM (default), Cognito or None.",
      "default": "IAM",
      "enum": ["IAM", "Cognito", "None"],
      "x-prompt": "How would you like users to authenticate with your API?",
      "x-priority": "important"
    },
    "directory": {
      "type": "string",
      "description": "The directory to store the application in.",
      "default": "packages",
      "x-prompt": "Which directory do you want to create the project in?",
      "x-priority": "important"
    },
    "moduleName": {
      "type": "string",
      "description": "Python module name"
    }
  },
  "required": ["name"]
}
