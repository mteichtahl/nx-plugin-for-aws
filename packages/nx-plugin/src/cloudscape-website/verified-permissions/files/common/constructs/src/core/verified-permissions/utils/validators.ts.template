/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { ICustomAttribute, UserPool } from 'aws-cdk-lib/aws-cognito';

/**
 * Options for validating schema attributes against a Cognito User Pool
 */
export interface SchemaValidationOptions {
  /** The Cognito User Pool to check for attribute definitions */
  userPool: UserPool;
  /** The required attributes that should be defined in the user pool */
  requiredAttributes: {
    [key: string]: ICustomAttribute;
  };
}

/**
 * Validates that all required attributes are properly defined in the Cognito User Pool schema.
 *
 * This ensures that the attributes used in Cedar policies are actually available in Cognito,
 * preventing runtime errors when attributes are referenced but don't exist in the identity source.
 *
 * The validation checks:
 * - Presence of the attributes in the options
 * - Proper configuration of each attribute
 *
 * @param options - The validation options containing the user pool and required attributes
 * @throws Error if any required attribute is not properly defined in the user pool
 */
export function validateSchemaAttributes(options: SchemaValidationOptions): void {
  const { userPool, requiredAttributes } = options;

  if (!userPool) {
    throw new Error('User pool must be provided for attribute validation');
  }

  if (!requiredAttributes || Object.keys(requiredAttributes).length === 0) {
    return; // No required attributes to validate
  }

  // Extract the user pool's schema attributes for reference checking
  // Note: CDK doesn't expose the schema property directly, so we warn about attributes
  // that might be missing but can't validate them statically

  // Instead, we encourage proper configuration by warning if attributes aren't defined through ICustomAttribute
  const attributeNames = Object.keys(requiredAttributes);

  for (const attributeName of attributeNames) {
    const attribute = requiredAttributes[attributeName];

    // Verify that the attribute is properly configured
    if (!attribute) {
      throw new Error(`Required attribute ${attributeName} is not properly defined`);
    }

    // Additional validations could go here, but we're limited by CDK's public APIs
    // For now, we just ensure the attribute exists in the required attributes collection
  }
}
