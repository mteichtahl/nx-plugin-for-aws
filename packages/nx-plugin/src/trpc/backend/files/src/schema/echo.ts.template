import { z } from 'zod';

export const EchoInputSchema = z.object({
  message: z.string(),
});

export type IEchoInput = z.TypeOf<typeof EchoInputSchema>;

export const EchoOutputSchema = z.object({
  result: z.string(),
});

export type IEchoOutput = z.TypeOf<typeof EchoOutputSchema>;
