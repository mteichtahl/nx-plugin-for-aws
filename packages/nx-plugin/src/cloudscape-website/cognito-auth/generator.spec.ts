/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree, updateJson } from '@nx/devkit';
import {
  COGNITO_AUTH_GENERATOR_INFO,
  tsCloudScapeWebsiteAuthGenerator,
} from './generator';
import { TsCloudScapeWebsiteAuthGeneratorSchema } from './schema';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
import { sharedConstructsGenerator } from '../../utils/shared-constructs';
import { expectHasMetricTags } from '../../utils/metrics.spec';

describe('cognito-auth generator', () => {
  let tree: Tree;

  const options: TsCloudScapeWebsiteAuthGeneratorSchema = {
    project: 'test-project',
    cognitoDomain: 'test',
    allowSignup: true,
  };

  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
    // Set up a mock project structure
    tree.write(
      'packages/test-project/project.json',
      JSON.stringify({
        name: 'test-project',
        sourceRoot: 'packages/test-project/src',
      }),
    );
  });

  it('should generate files', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      import { RuntimeConfigProvider } from './components/RuntimeConfig';
      import { RouterProvider, createRouter } from '@tanstack/react-router';

      export function Main() {

      const App = () => <RouterProvider router={router} />;

        return (
          <RuntimeConfigProvider>
            <App/>
          </RuntimeConfigProvider>
        );
      }
    `,
    );

    await tsCloudScapeWebsiteAuthGenerator(tree, options);

    // Verify component files are generated
    expect(
      tree.exists('packages/test-project/src/components/CognitoAuth/index.tsx'),
    ).toBeTruthy();

    // Verify shared constructs files are generated
    expect(
      tree.exists('packages/common/constructs/src/core/user-identity.ts'),
    ).toBeTruthy();
    expect(
      tree.exists('packages/common/constructs/src/core/index.ts'),
    ).toBeTruthy();

    // Create snapshot of the generated files
    expect(
      tree
        .read('packages/test-project/src/components/CognitoAuth/index.tsx')
        .toString(),
    ).toMatchSnapshot('cognito-auth-component');
    expect(
      tree.read('packages/common/constructs/src/core/index.ts').toString(),
    ).toMatchSnapshot('identity-index');
    expect(
      tree
        .read('packages/common/constructs/src/core/user-identity.ts')
        .toString(),
    ).toMatchSnapshot('user-identity');
  });

  it('should update main.tsx when RuntimeConfigProvider exists', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      import { RuntimeConfigProvider } from './components/RuntimeConfig';
      import { RouterProvider, createRouter } from '@tanstack/react-router';

      export function Main() {

        const App = () => <RouterProvider router={router} />;
        
        return (
          <RuntimeConfigProvider>
            <App/>
          </RuntimeConfigProvider>
        );
      }
    `,
    );

    await tsCloudScapeWebsiteAuthGenerator(tree, options);

    const mainTsxContent = tree
      .read('packages/test-project/src/main.tsx')
      .toString();
    // Verify CognitoAuth import is added
    expect(mainTsxContent).toContain(
      "import CognitoAuth from './components/CognitoAuth'",
    );
    // Verify CognitoAuth component is wrapped around children
    expect(mainTsxContent).toContain('<CognitoAuth>');
    expect(mainTsxContent).toContain('</CognitoAuth>');
    // Create snapshot of the modified main.tsx
    expect(mainTsxContent).toMatchSnapshot('main-tsx-with-runtime-config');
  });

  it('should handle main.tsx without RuntimeConfigProvider', async () => {
    // Setup main.tsx without RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      export function App() {
        return (
          <div>Hello World</div>
        );
      }
    `,
    );
    await expect(
      async () => await tsCloudScapeWebsiteAuthGenerator(tree, options),
    ).rejects.toThrowError();
  });

  it('should handle missing main.tsx', async () => {
    await expect(
      async () => await tsCloudScapeWebsiteAuthGenerator(tree, options),
    ).rejects.toThrowError();
  });

  it('should update shared constructs index.ts', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      import { RuntimeConfigProvider } from './components/RuntimeConfig';
      import { RouterProvider, createRouter } from '@tanstack/react-router';

      export function Main() {

        const App = () => <RouterProvider router={router} />;
        
        return (
          <RuntimeConfigProvider>
            <App/>
          </RuntimeConfigProvider>
        );
      }
    `,
    );
    // Setup initial shared constructs index
    tree.write(
      'packages/common/constructs/src/index.ts',
      'export const dummy = true;',
    );
    await tsCloudScapeWebsiteAuthGenerator(tree, options);
    const indexContent = tree
      .read('packages/common/constructs/src/core/index.ts')
      .toString();
    // Verify identity export is added
    expect(indexContent).toContain("export * from './user-identity.js'");
    // Create snapshot of the modified index
    expect(indexContent).toMatchSnapshot('common/constructs-index');
  });

  it('should add required dependencies', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      import { RuntimeConfigProvider } from './components/RuntimeConfig';
      import { RouterProvider, createRouter } from '@tanstack/react-router';

      export function App() {
        const App = () => <RouterProvider router={router} />;
        
        return (
          <RuntimeConfigProvider>
            <App/>
          </RuntimeConfigProvider>
        );
      }
    `,
    );
    await tsCloudScapeWebsiteAuthGenerator(tree, options);
    // Read package.json
    const packageJson = JSON.parse(tree.read('package.json').toString());
    // Verify dependencies are added
    expect(packageJson.dependencies).toMatchObject({
      constructs: expect.any(String),
      'aws-cdk-lib': expect.any(String),
      'oidc-client-ts': expect.any(String),
      'react-oidc-context': expect.any(String),
    });
  });

  it('should not be able to run the generator multiple times', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      import { RuntimeConfigProvider } from './components/RuntimeConfig';
      import { RouterProvider, createRouter } from '@tanstack/react-router';

      export function App() {

        const App = () => <RouterProvider router={router} />;

        return (
          <RuntimeConfigProvider>
            <App/>
          </RuntimeConfigProvider>
        );
      }
    `,
    );
    // First run to create files
    await tsCloudScapeWebsiteAuthGenerator(tree, options);
    // Run generator again
    await expect(
      async () => await tsCloudScapeWebsiteAuthGenerator(tree, options),
    ).rejects.toThrowErrorMatchingSnapshot();
  });

  it('should update AppLayout', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      import { RuntimeConfigProvider } from './components/RuntimeConfig';
      import { RouterProvider, createRouter } from '@tanstack/react-router';
      

      export function App() {
        
        const App = () => <RouterProvider router={router} />;

        return (
          <RuntimeConfigProvider>
            <App/>
          </RuntimeConfigProvider>
        );
      }
    `,
    );

    // Setup AppLayout.tsx with a basic component
    tree.write(
      'packages/test-project/src/components/AppLayout/index.tsx',
      `
      import * as React from 'react';
import { createContext, useCallback, useEffect, useState } from 'react';
import { NavItems } from './navitems';
import Config from '../../config';

import {
  BreadcrumbGroup,
  BreadcrumbGroupProps,
  SideNavigation,
  TopNavigation,
} from '@cloudscape-design/components';

import CloudscapeAppLayout, {
  AppLayoutProps,
} from '@cloudscape-design/components/app-layout';

import { matchByPath, useLocation, useNavigate } from '@tanstack/react-router';
import { Outlet } from '@tanstack/react-router';

const getBreadcrumbs = (
  pathName: string,
  search: string,
  defaultBreadcrumb: string,
  availableRoutes?: string[],
) => {
  const segments = [
    defaultBreadcrumb,
    ...pathName.split('/').filter((segment) => segment !== ''),
  ];

  return segments.map((segment, i) => {
    const href =
      i === 0
        ? '/'
        : \`/\${segments
            .slice(1, i + 1)
            .join('/')
            .replace('//', '/')}\`;

    const matched =
      !availableRoutes || availableRoutes.find((r) => matchByPath(r, href, {}));

    return {
      href: matched ? \`\${href}\${search}\` : '#',
      text: segment,
    };
  });
};

export interface AppLayoutContext {
  appLayoutProps: AppLayoutProps;
  setAppLayoutProps: (props: AppLayoutProps) => void;
  displayHelpPanel: (helpContent: React.ReactNode) => void;
}

/**
 * Context for updating/retrieving the AppLayout.
 */
export const AppLayoutContext = createContext({
  appLayoutProps: {},
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  setAppLayoutProps: (_: AppLayoutProps) => {},
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  displayHelpPanel: (_: React.ReactNode) => {},
});

/**
 * Defines the App layout and contains logic for routing.
 */
const AppLayout: React.FC = () => {
  const navigate = useNavigate();
  const appLayout = React.useRef<AppLayoutProps.Ref>(null);
  const [activeBreadcrumbs, setActiveBreadcrumbs] = useState<
    BreadcrumbGroupProps.Item[]
  >([{ text: '/', href: '/' }]);
  const [appLayoutProps, setAppLayoutProps] = useState<AppLayoutProps>({});
  const { pathname, search } = useLocation();

  const setAppLayoutPropsSafe = useCallback(
    (props: AppLayoutProps) => {
      JSON.stringify(appLayoutProps) !== JSON.stringify(props) &&
        setAppLayoutProps(props);
    },
    [appLayoutProps],
  );

  useEffect(() => {
    const breadcrumbs = getBreadcrumbs(
      pathname,
      Object.entries(search).reduce((p, [k, v]) => p + \`\${k}=\${v}\`, ''),
      '/',
    );
    setActiveBreadcrumbs(breadcrumbs);
  }, [pathname, search]);

  const onNavigate = useCallback(
    (e: CustomEvent<{ href: string; external?: boolean }>) => {
      if (!e.detail.external) {
        e.preventDefault();
        setAppLayoutPropsSafe({
          contentType: undefined,
        });
        navigate({ to: e.detail.href });
      }
    },
    [navigate, setAppLayoutPropsSafe],
  );

  return (
    <AppLayoutContext.Provider
      value={{
        appLayoutProps,
        setAppLayoutProps: setAppLayoutPropsSafe,
        displayHelpPanel: (helpContent: React.ReactNode) => {
          setAppLayoutPropsSafe({ tools: helpContent, toolsHide: false });
          appLayout.current?.openTools();
        },
      }}
    >
      <TopNavigation
        identity={{
          href: '/',
          title: Config.applicationName,
          logo: {
            src: Config.logo,
          },
        }}
      />
      <CloudscapeAppLayout
        ref={appLayout}
        breadcrumbs={
          <BreadcrumbGroup onFollow={onNavigate} items={activeBreadcrumbs} />
        }
        toolsHide
        navigation={
          <SideNavigation
            header={{ text: Config.applicationName, href: '/' }}
            activeHref={pathname}
            onFollow={onNavigate}
            items={NavItems}
          />
        }
        content={<Outlet />}
        {...appLayoutProps}
      />
    </AppLayoutContext.Provider>
  );
};

export default AppLayout;

    `,
    );

    await tsCloudScapeWebsiteAuthGenerator(tree, options);

    const appLayoutContent = tree
      .read('packages/test-project/src/components/AppLayout/index.tsx')
      .toString();

    // Verify useAuth import is added
    expect(appLayoutContent).toContain(
      "import { useAuth } from 'react-oidc-context'",
    );

    // Verify useAuth hook is used in the component
    expect(appLayoutContent).toContain(
      'const { user, removeUser, signoutRedirect, clearStaleState } = useAuth()',
    );

    // Verify TopNavigation has utilities attribute
    expect(appLayoutContent).toContain('utilities={[');
    expect(appLayoutContent).toContain("type: 'menu-dropdown'");
    expect(appLayoutContent).toContain(
      "text: `${user?.profile?.['cognito:username']}`",
    );
    expect(appLayoutContent).toContain("iconName: 'user-profile-active'");

    // Verify sign out functionality
    expect(appLayoutContent).toContain("id: 'signout'");
    expect(appLayoutContent).toContain("text: 'Sign out'");
    expect(appLayoutContent).toContain('removeUser()');
    expect(appLayoutContent).toContain('signoutRedirect(');
    expect(appLayoutContent).toContain('clearStaleState()');

    // Create snapshot of the modified AppLayout.tsx
    expect(appLayoutContent).toMatchSnapshot('app-layout-with-auth');
  });

  it('should allow an unqualified name to be specified', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      import { RuntimeConfigProvider } from './components/RuntimeConfig';
      import { RouterProvider, createRouter } from '@tanstack/react-router';

      export function App() {

        const App = () => <RouterProvider router={router} />;

        return (
          <RuntimeConfigProvider>
            <App/>
          </RuntimeConfigProvider>
        );
      }
    `,
    );

    // Use a fully-qualified name
    updateJson(tree, 'package.json', (packageJson) => ({
      ...packageJson,
      name: '@scope/source',
    }));
    tree.write(
      'packages/test-project/project.json',
      JSON.stringify({
        name: '@scope/test-project',
        sourceRoot: 'packages/test-project/src',
      }),
    );

    await tsCloudScapeWebsiteAuthGenerator(tree, {
      ...options,
      project: 'test-project', // unqualified
    });
  });

  it('should add generator metric to app.ts', async () => {
    // Set up test tree with shared constructs
    await sharedConstructsGenerator(tree);

    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      import { RuntimeConfigProvider } from './components/RuntimeConfig';
      import { RouterProvider, createRouter } from '@tanstack/react-router';

      export function App() {

      const App = () => <RouterProvider router={router} />;

        return (
          <RuntimeConfigProvider>
            <App/>
          </RuntimeConfigProvider>
        );
      }
    `,
    );

    // Call the generator function
    await tsCloudScapeWebsiteAuthGenerator(tree, options);

    // Verify the metric was added to app.ts
    expectHasMetricTags(tree, COGNITO_AUTH_GENERATOR_INFO.metric);
  });
});
