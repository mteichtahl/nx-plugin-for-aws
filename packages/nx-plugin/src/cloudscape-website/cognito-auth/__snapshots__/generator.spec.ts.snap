// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cognito-auth generator > should generate files > cognito-auth-component 1`] = `
"import React, { PropsWithChildren, useEffect } from 'react';
import { AuthProvider, AuthProviderProps, useAuth } from 'react-oidc-context';
import { Alert, Spinner } from '@cloudscape-design/components';
import { useRuntimeConfig } from '../../hooks/useRuntimeConfig';

/**
 * Sets up the Cognito auth.
 *
 * This assumes a runtime-config.json file is present at '/'. In order for Auth to be set up automatically,
 * the runtime-config.json must have the cognitoProps set.
 */
const CognitoAuth: React.FC<PropsWithChildren> = ({ children }) => {
  const { cognitoProps } = useRuntimeConfig();

  if (!cognitoProps) {
    if (import.meta.env.MODE === 'serve-local') {
      // In serve-local mode with no cognitoProps available, we skip login
      return <AuthProvider>{children}</AuthProvider>;
    }
    return (
      <Alert type="error" header="Runtime config configuration error">
        The cognitoProps have not been configured in the runtime-config.json.
      </Alert>
    );
  }

  const cognitoAuthConfig: AuthProviderProps = {
    authority: \`https://cognito-idp.\${cognitoProps.region}.amazonaws.com/\${cognitoProps.userPoolId}\`,
    client_id: cognitoProps.userPoolWebClientId,
    redirect_uri: window.location.origin,
    response_type: 'code',
    scope: 'email openid profile',
  };

  return (
    <AuthProvider {...cognitoAuthConfig}>
      <CognitoAuthInternal>{children}</CognitoAuthInternal>
    </AuthProvider>
  );
};

const CognitoAuthInternal: React.FC<PropsWithChildren> = ({ children }) => {
  const auth = useAuth();

  useEffect(() => {
    if (!auth.isAuthenticated && !auth.isLoading) {
      auth.signinRedirect();
    }
  }, [auth]);

  if (auth.isAuthenticated) {
    return children;
  } else if (auth.error) {
    return (
      <Alert type="error" header="Configuration error">
        Error contacting Cognito. Please check your runtime-config.json is
        configured with the correct endpoints.
      </Alert>
    );
  } else {
    return <Spinner />;
  }
};

export default CognitoAuth;
"
`;

exports[`cognito-auth generator > should generate files > identity-index 1`] = `
"export * from './user-identity.js';
export * from './app.js';
export * from './runtime-config.js';
"
`;

exports[`cognito-auth generator > should generate files > user-identity 1`] = `
"import {
  IdentityPool,
  UserPoolAuthenticationProvider,
} from 'aws-cdk-lib/aws-cognito-identitypool';
import { CfnOutput, Duration, Lazy, Stack } from 'aws-cdk-lib';
import {
  AccountRecovery,
  CfnManagedLoginBranding,
  CfnUserPoolDomain,
  FeaturePlan,
  Mfa
  OAuthScope,
  UserPool,
  UserPoolClient,
} from 'aws-cdk-lib/aws-cognito';
import { IRole } from 'aws-cdk-lib/aws-iam';
import { Construct } from 'constructs';
import { RuntimeConfig } from './runtime-config.js';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';

const WEB_CLIENT_ID = 'WebClient';
/**
 * Creates a UserPool and Identity Pool with sane defaults configured intended for usage from a web client.
 */
export class UserIdentity extends Construct {
  public readonly region: string;
  public readonly identityPool: IdentityPool;
  public readonly userPool: UserPool;
  public readonly userPoolClient: UserPoolClient;
  public readonly userPoolClients: UserPoolClient[];
  public readonly userPoolDomain: CfnUserPoolDomain;
  public readonly authenticatedRole: IRole;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.region = Stack.of(this).region;
    this.userPool = this.createUserPool();
    this.userPoolDomain = this.createUserPoolDomain(this.userPool);
    this.userPoolClient = this.createUserPoolClient(this.userPool);
    this.userPoolClients = [this.userPoolClient];
    this.identityPool = this.createIdentityPool(
      this.userPool,
      this.userPoolClient
    );
    this.authenticatedRole = this.identityPool.authenticatedRole;
    this.createManagedLoginBranding(
      this.userPool,
      this.userPoolClient,
      this.userPoolDomain
    );

    RuntimeConfig.ensure(this).config.cognitoProps = {
      region: Stack.of(this).region,
      identityPoolId: this.identityPool.identityPoolId,
      userPoolId: this.userPool.userPoolId,
      userPoolWebClientId: this.userPoolClient.userPoolClientId,
    };

    new CfnOutput(this, \`\${id}-UserPoolId\`, {
      value: this.userPool.userPoolId,
    });

    new CfnOutput(this, \`\${id}-IdentityPoolId\`, {
      value: this.identityPool.identityPoolId,
    });
  }

  private createUserPool = () =>
    new UserPool(this, 'UserPool', {
      deletionProtection: true,
      passwordPolicy: {
        minLength: 8,
        requireLowercase: true,
        requireUppercase: true,
        requireDigits: true,
        requireSymbols: true,
        tempPasswordValidity: Duration.days(3),
      },
      mfa: Mfa.REQUIRED,
      featurePlan: FeaturePlan.ESSENTIALS,
      mfaSecondFactor: { sms: true, otp: true },
      signInCaseSensitive: false,
      signInAliases: { username: true, email: true },
      accountRecovery: AccountRecovery.EMAIL_ONLY,
      selfSignUpEnabled: true,
      standardAttributes: {
        phoneNumber: { required: false },
        email: { required: true },
        givenName: { required: true },
        familyName: { required: true },
      },
      autoVerify: {
        email: true,
        phone: true,
      },
      keepOriginal: {
        email: true,
        phone: true,
      },
    });

  private createUserPoolDomain = (userPool: UserPool) =>
    new CfnUserPoolDomain(this, 'UserPoolDomain', {
      domain: \`test-\${Stack.of(this).account}\`,
      userPoolId: userPool.userPoolId,
      managedLoginVersion: 2,
    });

  private createUserPoolClient = (userPool: UserPool) => {
    const lazilyComputedCallbackUrls = Lazy.list({
      produce: () =>
        [
          'http://localhost:4200',
          'http://localhost:4300',
          \`https://\${Stack.of(this).region}.console.aws.amazon.com\`,
        ].concat(
          this.findCloudFrontDistributions().map(
            (d) => \`https://\${d.domainName}\`
          )
        ),
    });

    return userPool.addClient(WEB_CLIENT_ID, {
      authFlows: {
        userPassword: true,
        userSrp: true,
        user: true,
      },
      oAuth: {
        flows: {
          authorizationCodeGrant: true,
        },
        scopes: [OAuthScope.EMAIL, OAuthScope.OPENID, OAuthScope.PROFILE],
        callbackUrls: lazilyComputedCallbackUrls,
        logoutUrls: lazilyComputedCallbackUrls,
      },
      preventUserExistenceErrors: true,
    });
  };

  private createIdentityPool = (
    userPool: UserPool,
    userPoolClient: UserPoolClient
  ) => {
    const identityPool = new IdentityPool(this, 'IdentityPool');

    identityPool.addUserPoolAuthentication(
      new UserPoolAuthenticationProvider({
        userPool,
        userPoolClient,
      })
    );

    return identityPool;
  };

  private createManagedLoginBranding = (
    userPool: UserPool,
    userPoolClient: UserPoolClient,
    userPoolDomain: CfnUserPoolDomain
  ) => {
    new CfnManagedLoginBranding(this, 'ManagedLoginBranding', {
      userPoolId: userPool.userPoolId,
      clientId: userPoolClient.userPoolClientId,
      useCognitoProvidedValues: true,
    }).node.addDependency(userPoolClient, userPool, userPoolDomain);
  };

  private findCloudFrontDistributions = (): Distribution[] =>
    Stack.of(this)
      .node.findAll()
      .filter((child) => child instanceof Distribution);
}
"
`;

exports[`cognito-auth generator > should not be able to run the generator multiple times 1`] = `[Error: This generator has already been run on test-project.]`;

exports[`cognito-auth generator > should update AppLayout > app-layout-with-auth 1`] = `
"import { useAuth } from 'react-oidc-context';
import * as React from 'react';
import { createContext, useCallback, useEffect, useState } from 'react';
import { NavItems } from './navitems';
import Config from '../../config';

import {
  BreadcrumbGroup,
  BreadcrumbGroupProps,
  SideNavigation,
  TopNavigation,
} from '@cloudscape-design/components';

import CloudscapeAppLayout, {
  AppLayoutProps,
} from '@cloudscape-design/components/app-layout';

import { matchByPath, useLocation, useNavigate } from '@tanstack/react-router';
import { Outlet } from '@tanstack/react-router';

const getBreadcrumbs = (
  pathName: string,
  search: string,
  defaultBreadcrumb: string,
  availableRoutes?: string[],
) => {
  const segments = [
    defaultBreadcrumb,
    ...pathName.split('/').filter((segment) => segment !== ''),
  ];

  return segments.map((segment, i) => {
    const href =
      i === 0
        ? '/'
        : \`/\${segments
            .slice(1, i + 1)
            .join('/')
            .replace('//', '/')}\`;

    const matched =
      !availableRoutes || availableRoutes.find((r) => matchByPath(r, href, {}));

    return {
      href: matched ? \`\${href}\${search}\` : '#',
      text: segment,
    };
  });
};

export interface AppLayoutContext {
  appLayoutProps: AppLayoutProps;
  setAppLayoutProps: (props: AppLayoutProps) => void;
  displayHelpPanel: (helpContent: React.ReactNode) => void;
}

/**
 * Context for updating/retrieving the AppLayout.
 */
export const AppLayoutContext = createContext({
  appLayoutProps: {},
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  setAppLayoutProps: (_: AppLayoutProps) => {},
  // eslint-disable-next-line @typescript-eslint/no-empty-function
  displayHelpPanel: (_: React.ReactNode) => {},
});

/**
 * Defines the App layout and contains logic for routing.
 */
const AppLayout: React.FC = () => {
  const { user, removeUser, signoutRedirect, clearStaleState } = useAuth();
  const navigate = useNavigate();
  const appLayout = React.useRef<AppLayoutProps.Ref>(null);
  const [activeBreadcrumbs, setActiveBreadcrumbs] = useState<
    BreadcrumbGroupProps.Item[]
  >([{ text: '/', href: '/' }]);
  const [appLayoutProps, setAppLayoutProps] = useState<AppLayoutProps>({});
  const { pathname, search } = useLocation();
  const setAppLayoutPropsSafe = useCallback(
    (props: AppLayoutProps) => {
      JSON.stringify(appLayoutProps) !== JSON.stringify(props) &&
        setAppLayoutProps(props);
    },
    [appLayoutProps],
  );
  useEffect(() => {
    const breadcrumbs = getBreadcrumbs(
      pathname,
      Object.entries(search).reduce((p, [k, v]) => p + \`\${k}=\${v}\`, ''),
      '/',
    );
    setActiveBreadcrumbs(breadcrumbs);
  }, [pathname, search]);
  const onNavigate = useCallback(
    (
      e: CustomEvent<{
        href: string;
        external?: boolean;
      }>,
    ) => {
      if (!e.detail.external) {
        e.preventDefault();
        setAppLayoutPropsSafe({
          contentType: undefined,
        });
        navigate({ to: e.detail.href });
      }
    },
    [navigate, setAppLayoutPropsSafe],
  );
  return (
    <AppLayoutContext.Provider
      value={{
        appLayoutProps,
        setAppLayoutProps: setAppLayoutPropsSafe,
        displayHelpPanel: (helpContent: React.ReactNode) => {
          setAppLayoutPropsSafe({ tools: helpContent, toolsHide: false });
          appLayout.current?.openTools();
        },
      }}
    >
      <TopNavigation
        identity={{
          href: '/',
          title: Config.applicationName,
          logo: {
            src: Config.logo,
          },
        }}
        utilities={[
          {
            type: 'menu-dropdown',
            text: \`\${user?.profile?.['cognito:username']}\`,
            iconName: 'user-profile-active',
            onItemClick: (e) => {
              if (e.detail.id === 'signout') {
                removeUser();
                signoutRedirect({
                  post_logout_redirect_uri: window.location.origin,
                  extraQueryParams: {
                    redirect_uri: window.location.origin,
                    response_type: 'code',
                  },
                });
                clearStaleState();
              }
            },
            items: [{ id: 'signout', text: 'Sign out' }],
          },
        ]}
      />
      <CloudscapeAppLayout
        ref={appLayout}
        breadcrumbs={
          <BreadcrumbGroup onFollow={onNavigate} items={activeBreadcrumbs} />
        }
        toolsHide
        navigation={
          <SideNavigation
            header={{ text: Config.applicationName, href: '/' }}
            activeHref={pathname}
            onFollow={onNavigate}
            items={NavItems}
          />
        }
        content={<Outlet />}
        {...appLayoutProps}
      />
    </AppLayoutContext.Provider>
  );
};

export default AppLayout;
"
`;

exports[`cognito-auth generator > should update main.tsx when RuntimeConfigProvider exists > main-tsx-with-runtime-config 1`] = `
"import { useAuth } from 'react-oidc-context';
import CognitoAuth from './components/CognitoAuth';
import { useRuntimeConfig } from './hooks/useRuntimeConfig';
import RuntimeConfigProvider from './components/RuntimeConfig';
import { RuntimeConfigProvider } from './components/RuntimeConfig';
import { RouterProvider, createRouter } from '@tanstack/react-router';

export function Main() {
  const App = () => <RouterProvider router={router} />;

  return (
    <RuntimeConfigProvider>
      <CognitoAuth>
        <RuntimeConfigProvider>
          <App />
        </RuntimeConfigProvider>
      </CognitoAuth>
    </RuntimeConfigProvider>
  );
}
"
`;

exports[`cognito-auth generator > should update shared constructs index.ts > common/constructs-index 1`] = `
"export * from './user-identity.js';
export * from './app.js';
export * from './runtime-config.js';
"
`;
